py/private/py_pex_binary.bzl

```diff
@@ -0,0 +1,119 @@
+"Create python zip file https://peps.python.org/pep-0441/ (PEX)"
+
+load("@rules_python//python:defs.bzl", "PyInfo")
+load("//py/private:py_semantics.bzl", _py_semantics = "semantics")
+load("//py/private/toolchain:types.bzl", "PY_TOOLCHAIN")
+
+def _runfiles_path(file, workspace):
+    if file.short_path.startswith("../"):
+        return file.short_path[3:]
+    else:
+        return workspace + "/" + file.short_path
+
+exclude_paths = [
+    "toolchain",
+    "aspect_rules_py/py/tools/",
+    "rules_python~",
+    "aspect_rules_py~/py/tools/"
+]
+
+def _map_srcs(f, workspace):
+    dest_path = _runfiles_path(f, workspace)
+
+    # TODO: better way to exclude hermetic toolchain.
+    for exclude in exclude_paths:
+        if dest_path.find(exclude) != -1:
+            return []
+
+
+    site_packages_i = f.path.find("site-packages")
+
+    # determine if the src if a third party.
+    if site_packages_i != -1 and f.path.count("/", site_packages_i) == 2:
+        if f.path.find("dist-info", site_packages_i) != -1:
+            return ["--distinfo", f.dirname]
+        return ["--dep", f.dirname]
+
+    elif site_packages_i == -1:
+        return ["--source=%s=%s" % (f.path, dest_path)]
+
+    return []
+
+def _py_python_pex_impl(ctx):
+    py_toolchain = _py_semantics.resolve_toolchain(ctx)
+
+    binary = ctx.attr.binary
+    runfiles = binary[DefaultInfo].data_runfiles
+
+    output = ctx.actions.declare_file(ctx.attr.name + ".pex")
+
+    args = ctx.actions.args()
+
+    # Copy workspace name here to prevent ctx
+    # being transferred to the execution phase.
+    workspace_name = str(ctx.workspace_name)
+
+    args.add_all(
+        ctx.attr.inject_env.items(), 
+        map_each = lambda e: "--inject-env=%s=%s" % (e[0], e[1]),
+        allow_closure = True,
```

Comment:
Does this need to capture the closure? Seems like it only references the var in the lambda.

Reply:
nope, left over. will remove

Reply:
Please see the comment below the argument. 

https://github.com/aspect-build/rules_py/pull/324/files#diff-bdbfb557a8b870d00314cf49db0fad2611b37d9d20d6c055034aef05b30a6e65R77

Reply:
> # this is needed to allow passing a lambda (with workspace_name) to map_each

Reply:
But this isn't that call, the `map_each` here is `lambda e: "--inject-env=%s=%s" % (e[0], e[1]),` so how is the call below related?

Reply:
Lambda for `inject_env`'s `map_each` does not contain any references to the `workspace_name`.

Reply:
i could get rid of it by creating a global `def`, this feels more localized and i did not see any performance degradation.

Reply:
https://bazel.build/rules/lib/builtins/Args#parameters_1 says
> To avoid unintended retention of large analysis-phase data structures into the execution phase, the map_each function must be declared by a top-level def statement; [...]

so I think it's better to rewrite it as a global `def`

Reply:
In that case, can the comment be updated, as it's unrelated to what is happening in his call.

Reply:
fixed the comment.